lib/src/rules/unreachable_from_main.dart

```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule);
+
+  final LintRule rule;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    // TODO(a14n): add support of libs with parts
+    if (node.directives.whereType<PartOfDirective>().isNotEmpty) return;
+    if (node.directives.whereType<PartDirective>().isNotEmpty) return;
+
+    var topDeclarations = node.declarations
+        .expand((e) => [
+              if (e is TopLevelVariableDeclaration)
+                ...e.variables.variables
+              else
+                e,
+            ])
+        .toSet();
+
+    var entryPoints = topDeclarations.where(_isEntryPoint).toList();
+    if (entryPoints.isEmpty) return;
+
+    var declarationByElement = <Element, Declaration>{};
+    for (var declaration in topDeclarations) {
+      var element = declaration.declaredElement;
+      if (element != null) {
+        if (element is TopLevelVariableElement) {
+          declarationByElement[element] = declaration;
+          var getter = element.getter;
+          if (getter != null) declarationByElement[getter] = declaration;
+          var setter = element.setter;
+          if (setter != null) declarationByElement[setter] = declaration;
+        } else {
+          declarationByElement[element] = declaration;
+        }
+      }
+    }
+
+    var dependencies = Map<Declaration, Set<Declaration>>.fromIterable(
+      topDeclarations,
+      value: (declaration) =>
+          DartTypeUtilities.traverseNodesInDFS(declaration as Declaration)
+              .expand((e) => [
+                    if (e is SimpleIdentifier) e.staticElement,
+                    // with `id++` staticElement of `id` is null
+                    if (e is CompoundAssignmentExpression) ...[
+                      e.readElement,
+                      e.writeElement,
+                    ],
+                  ])
+              .whereNotNull()
+              .map((e) {
+                while (e.enclosingElement2 != null &&
+                    e.enclosingElement2 is! CompilationUnitElement) {
+                  e = e.enclosingElement2!;
+                }
+                return e;
+              })
+              .map((e) => declarationByElement[e])
+              .whereNotNull()
+              .where((e) => e != declaration)
+              .toSet(),
+    );
+
+    var usedMembers = entryPoints.toSet();
+    var toTraverse = Queue.from(usedMembers);
+    while (toTraverse.isNotEmpty) {
+      var declaration = toTraverse.removeLast();
+      for (var dep in dependencies[declaration]!) {
+        if (usedMembers.add(dep)) {
+          toTraverse.add(dep);
+        }
+      }
+    }
+
+    var unusedMembers = topDeclarations.difference(usedMembers).where((e) {
+      var element = e.declaredElement;
+      return element != null &&
+          element.isPublic &&
+          !element.hasVisibleForTesting;
+    });
+    unusedMembers.forEach(rule.reportLint);
```

Comment:
Given that these are declarations, it would be better if we could report the lint on the name of the declaration, or the `extension` keyword in the case of an unnamed extension.

Reply:
done